### PR TITLE
Cut v2.50.0-rc.0

### PR DESCRIPTION
**Pull request by Nexucis**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,38 @@
 # Changelog
 
+
+## 2.50.0-rc.0 / 2024-01-26
+
+* [FEATURE] Add new feature flag `created-timestamp-zero-ingestion`. #12733 #13279
+* [FEATURE] TSDB/{index,compact}: allow using custom postings encoding format. #13242 #13407 #13391
+* [FEATURE] Promtool: Add `analyze` histograms command #12331
+* [FEATURE] TSDB/compaction: Add a way to enable overlapping compaction #13282 #13393 #13398
+* [FEATURE] Add automatic memory limit handling. Activated using the flag `auto-gomemlimit.ratio` #13395
```
Comment (SuperQ) [inline CHANGELOG.md:10]:
```suggestion
* [FEATURE] Add automatic memory limit handling. Activated using the feature flag `auto-gomemlimit` #13395
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,38 @@
 # Changelog
 
+
+## 2.50.0-rc.0 / 2024-01-26
+
+* [FEATURE] Add new feature flag `created-timestamp-zero-ingestion`. #12733 #13279
```
Comment (ArthurSens) [inline CHANGELOG.md:6]:
```suggestion
* [FEATURE] Add support for ingesting zeros as created timestamps (enabled under the feature-flag `created-timestamp-zero-ingestion)`. #12733 #13279
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,37 @@
...
+* [ENHANCEMENT] Enable reusing memory when converting between histogram types. #13215
+* [ENHANCEMENT] PERFORMANCE: Reuse float histogram objects. #13276 #13289
+* [ENHANCEMENT] SD: Extend Kuma SD configuration to allow users to specify ClientId. #13278
+* [ENHANCEMENT] Optimize histogram iterators. #13340
+* [ENHANCEMENT] PROMQL: Use natural sort in `sort_by_label` and `sort_by_label_desc`. #13411
+* [ENHANCEMENT] Native Histograms: support `native_histogram_min_bucket_factor` in scrape_config. #13222
+* [ENHANCEMENT] Metrics: Add warnings for histogramRate applied with isCounter not matching counter/gauge histogram. #13392
+* [ENHANCEMENT] TSDB/PERFORMANCE: make transaction isolation data structures smaller. #13015
+* [ENHANCEMENT] TSDB/postings: use Loser Tree for merge. #12878
+* [ENHANCEMENT] TSDB/PERFORMANCE: simplify internal series delete function. #13261
```
Comment (roidelapluie) [inline CHANGELOG.md:21]:
```suggestion
* [ENHANCEMENT] TSDB: simplify internal series delete function. #13261
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,37 @@
...
+* [ENHANCEMENT] TSDB/AGENT: Performance improvement by making the global hash lookup table smaller. #13262
+* [ENHANCEMENT] Labels/PERFORMANCE: Add DropMetricName function, used in PromQL. #13446
+* [ENHANCEMENT] PERFORMANCE: Optimize label values with matchers by taking shortcuts. #13426
+* [ENHANCEMENT] SD/PERFORMANCE: check preconditions earlier and avoid unnecessary checks or iterations in kube_sd. #13408
```
Comment (roidelapluie) [inline CHANGELOG.md:25]:
```suggestion
* [ENHANCEMENT] Kubernetes SD: check preconditions earlier and avoid unnecessary checks or iterations in kube_sd. #13408
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,5 +1,37 @@
...
+* [ENHANCEMENT] PromQL: simplify Native Histogram arithmetics. #13267
+* [ENHANCEMENT] Promtool: Improve visibility for `promtool test rules` with JSON colored formatting. #13342
+* [ENHANCEMENT] CONSOLES: exclude iowait and steal from CPU Utilisation. #9593
```
Comment (roidelapluie) [inline CHANGELOG.md:28]:
```suggestion
* [ENHANCEMENT] Consoles: Exclude iowait and steal from CPU Utilisation. #9593
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,5 +1,37 @@
...
+* [BUGFIX] Fix quality value in HTTP Accept header. #13313
+* [BUGFIX] Fix reusing float histograms. #13289
+* [BUGFIX] UI: fix handling of scalar and string in isHeatmapData. #13371
+* [CHANGE] REMOTE_WRITE: Drop old inmemory samples. Activated using the config entry `sample_age_limit` #13002
```
Comment (roidelapluie) [inline CHANGELOG.md:32]:
```suggestion
* [CHANGE] Remote Write: Drop old inmemory samples. Activated using the config entry `sample_age_limit` #13002
```

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,5 +1,37 @@
...
+* [CHANGE] REMOTE_WRITE: Error `storage.ErrTooOldSample` is now generating HTTP error 400 instead of HTTP error 500. #13335
```
Comment (roidelapluie) [inline CHANGELOG.md:33]:
```suggestion
* [CHANGE] Remote Write: Error `storage.ErrTooOldSample` is now generating HTTP error 400 instead of HTTP error 500. #13335
```

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,5 +1,37 @@
...
+* [ENHANCEMENT] Optimize histogram iterators. #13340
+* [ENHANCEMENT] PROMQL: Use natural sort in `sort_by_label` and `sort_by_label_desc`. #13411
+* [ENHANCEMENT] Native Histograms: support `native_histogram_min_bucket_factor` in scrape_config. #13222
+* [ENHANCEMENT] Metrics: Add warnings for histogramRate applied with isCounter not matching counter/gauge histogram. #13392
+* [ENHANCEMENT] TSDB/PERFORMANCE: make transaction isolation data structures smaller. #13015
+* [ENHANCEMENT] TSDB/postings: use Loser Tree for merge. #12878
+* [ENHANCEMENT] TSDB/PERFORMANCE: simplify internal series delete function. #13261
+* [ENHANCEMENT] TSDB/AGENT: Performance improvement by making the global hash lookup table smaller. #13262
+* [ENHANCEMENT] Labels/PERFORMANCE: Add DropMetricName function, used in PromQL. #13446
+* [ENHANCEMENT] PERFORMANCE: Optimize label values with matchers by taking shortcuts. #13426
```
Comment (roidelapluie) [inline CHANGELOG.md:24]:
```suggestion
* [ENHANCEMENT] TSDB: Optimize label values with matchers by taking shortcuts. #13426
```

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,5 +1,37 @@
...
+* [ENHANCEMENT] SD: Extend Kuma SD configuration to allow users to specify ClientId. #13278
+* [ENHANCEMENT] Optimize histogram iterators. #13340
+* [ENHANCEMENT] PROMQL: Use natural sort in `sort_by_label` and `sort_by_label_desc`. #13411
+* [ENHANCEMENT] Native Histograms: support `native_histogram_min_bucket_factor` in scrape_config. #13222
+* [ENHANCEMENT] Metrics: Add warnings for histogramRate applied with isCounter not matching counter/gauge histogram. #13392
+* [ENHANCEMENT] TSDB/PERFORMANCE: make transaction isolation data structures smaller. #13015
+* [ENHANCEMENT] TSDB/postings: use Loser Tree for merge. #12878
+* [ENHANCEMENT] TSDB/PERFORMANCE: simplify internal series delete function. #13261
+* [ENHANCEMENT] TSDB/AGENT: Performance improvement by making the global hash lookup table smaller. #13262
+* [ENHANCEMENT] Labels/PERFORMANCE: Add DropMetricName function, used in PromQL. #13446
```
Comment (roidelapluie) [inline CHANGELOG.md:23]: we do not need to mention this

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,5 +1,37 @@
...
+* [ENHANCEMENT] PERFORMANCE: Reuse float histogram objects. #13276 #13289
+* [ENHANCEMENT] SD: Extend Kuma SD configuration to allow users to specify ClientId. #13278
+* [ENHANCEMENT] Optimize histogram iterators. #13340
+* [ENHANCEMENT] PROMQL: Use natural sort in `sort_by_label` and `sort_by_label_desc`. #13411
+* [ENHANCEMENT] Native Histograms: support `native_histogram_min_bucket_factor` in scrape_config. #13222
+* [ENHANCEMENT] Metrics: Add warnings for histogramRate applied with isCounter not matching counter/gauge histogram. #13392
+* [ENHANCEMENT] TSDB/PERFORMANCE: make transaction isolation data structures smaller. #13015
+* [ENHANCEMENT] TSDB/postings: use Loser Tree for merge. #12878
+* [ENHANCEMENT] TSDB/PERFORMANCE: simplify internal series delete function. #13261
+* [ENHANCEMENT] TSDB/AGENT: Performance improvement by making the global hash lookup table smaller. #13262
```
Comment (roidelapluie) [inline CHANGELOG.md:22]:
```suggestion
* [ENHANCEMENT] Agent: Performance improvement by making the global hash lookup table smaller. #13262
```

File: CHANGELOG.md [hunk 11]
```diff
@@ -1,5 +1,37 @@
...
+* [ENHANCEMENT] Promtool: allow setting multiple matchers to `promtool tsdb dump`. #13296
+* [ENHANCEMENT] PromQL: Restore more efficient version of `NewPossibleNonCounterInfo` annotation. #13022
+* [ENHANCEMENT] Enable reusing memory when converting between histogram types. #13215
+* [ENHANCEMENT] PERFORMANCE: Reuse float histogram objects. #13276 #13289
+* [ENHANCEMENT] SD: Extend Kuma SD configuration to allow users to specify ClientId. #13278
+* [ENHANCEMENT] Optimize histogram iterators. #13340
+* [ENHANCEMENT] PROMQL: Use natural sort in `sort_by_label` and `sort_by_label_desc`. #13411
+* [ENHANCEMENT] Native Histograms: support `native_histogram_min_bucket_factor` in scrape_config. #13222
+* [ENHANCEMENT] Metrics: Add warnings for histogramRate applied with isCounter not matching counter/gauge histogram. #13392
+* [ENHANCEMENT] TSDB/PERFORMANCE: make transaction isolation data structures smaller. #13015
```
Comment (roidelapluie) [inline CHANGELOG.md:19]:
```suggestion
* [ENHANCEMENT] TSDB: Make transaction isolation data structures smaller. #13015
```

File: CHANGELOG.md [hunk 12]
```diff
@@ -1,5 +1,37 @@
...
+* [FEATURE] Add support for ingesting zeros as created timestamps (enabled under the feature-flag `created-timestamp-zero-ingestion`). #12733 #13279
+* [FEATURE] TSDB/{index,compact}: allow using custom postings encoding format. #13242 #13407 #13391
+* [FEATURE] Promtool: Add `analyze` histograms command #12331
+* [FEATURE] TSDB/compaction: Add a way to enable overlapping compaction #13282 #13393 #13398
+* [FEATURE] Add automatic memory limit handling. Activated using the feature flag `auto-gomemlimit` #13395
+* [ENHANCEMENT] Promtool: allow setting multiple matchers to `promtool tsdb dump`. #13296
+* [ENHANCEMENT] PromQL: Restore more efficient version of `NewPossibleNonCounterInfo` annotation. #13022
+* [ENHANCEMENT] Enable reusing memory when converting between histogram types. #13215
+* [ENHANCEMENT] PERFORMANCE: Reuse float histogram objects. #13276 #13289
+* [ENHANCEMENT] SD: Extend Kuma SD configuration to allow users to specify ClientId. #13278
```
Comment (roidelapluie) [inline CHANGELOG.md:14]:
```suggestion
* [ENHANCEMENT] Kuma SD: Extend configuration to allow users to specify ClientId. #13278
```

File: CHANGELOG.md [hunk 13]
```diff
@@ -1,5 +1,37 @@
...
+* [ENHANCEMENT] Optimize histogram iterators. #13340
+* [ENHANCEMENT] PROMQL: Use natural sort in `sort_by_label` and `sort_by_label_desc`. #13411
```
[the remark this answers is not in the document]
Comment (bboreham) [reply CHANGELOG.md:16]: Suggest to add 'experimental' on this line.

File: CHANGELOG.md [hunk 14]
```diff
@@ -1,5 +1,37 @@
...
+* [ENHANCEMENT] Native Histograms: support `native_histogram_min_bucket_factor` in scrape_config. #13222
+* [ENHANCEMENT] Metrics: Add warnings for histogramRate applied with isCounter not matching counter/gauge histogram. #13392
+* [ENHANCEMENT] TSDB/PERFORMANCE: make transaction isolation data structures smaller. #13015
+* [ENHANCEMENT] TSDB/postings: use Loser Tree for merge. #12878
```
Comment (bboreham) [inline CHANGELOG.md:20]:
```suggestion
* [ENHANCEMENT] TSDB/postings: optimize merge using Loser Tree. #12878
```

File: CHANGELOG.md [hunk 15]
```diff
@@ -1,5 +1,37 @@
...
+* [ENHANCEMENT] TSDB/AGENT: Performance improvement by making the global hash lookup table smaller. #13262
+* [ENHANCEMENT] Labels/PERFORMANCE: Add DropMetricName function, used in PromQL. #13446
+* [ENHANCEMENT] PERFORMANCE: Optimize label values with matchers by taking shortcuts. #13426
+* [ENHANCEMENT] SD/PERFORMANCE: check preconditions earlier and avoid unnecessary checks or iterations in kube_sd. #13408
+* [ENHANCEMENT] PromQL: simplify Native Histogram arithmetics. #13267
+* [ENHANCEMENT] Promtool: Improve visibility for `promtool test rules` with JSON colored formatting. #13342
+* [ENHANCEMENT] CONSOLES: exclude iowait and steal from CPU Utilisation. #9593
+* [BUGFIX] Fix quality value in HTTP Accept header. #13313
+* [BUGFIX] Fix reusing float histograms. #13289
+* [BUGFIX] UI: fix handling of scalar and string in isHeatmapData. #13371
```
Comment (bboreham) [inline CHANGELOG.md:31]: Can we make this more end-user oriented?  When would they notice the difference?

File: CHANGELOG.md (hunk 6 above)
Comment (bboreham) [reply CHANGELOG.md:32]: Is this not an enhancement?  CHANGE to me means a breaking change I get without any config change.

File: CHANGELOG.md (hunk 7 above)
Comment (bboreham) [reply CHANGELOG.md:33]: I would put breaking changes at the top.

File: CHANGELOG.md (hunk 9 above)
Comment (bboreham) [reply CHANGELOG.md:23]: We could mention it as an optimization to PromQL expressions like `abs(foo)`.
It's also possibly a pessimization, since we no longer cache the output.  If people have labels like `__i_sort_before__name`.

File: CHANGELOG.md [hunk 16]
```diff
@@ -1,5 +1,37 @@
...
+* [ENHANCEMENT] Native Histograms: support `native_histogram_min_bucket_factor` in scrape_config. #13222
+* [ENHANCEMENT] Metrics: Add warnings for histogramRate applied with isCounter not matching counter/gauge histogram. #13392
+* [ENHANCEMENT] TSDB/PERFORMANCE: make transaction isolation data structures smaller. #13015
+* [ENHANCEMENT] TSDB/postings: use Loser Tree for merge. #12878
+* [ENHANCEMENT] TSDB/PERFORMANCE: simplify internal series delete function. #13261
+* [ENHANCEMENT] TSDB/AGENT: Performance improvement by making the global hash lookup table smaller. #13262
+* [ENHANCEMENT] Labels/PERFORMANCE: Add DropMetricName function, used in PromQL. #13446
+* [ENHANCEMENT] PERFORMANCE: Optimize label values with matchers by taking shortcuts. #13426
+* [ENHANCEMENT] SD/PERFORMANCE: check preconditions earlier and avoid unnecessary checks or iterations in kube_sd. #13408
+* [ENHANCEMENT] PromQL: simplify Native Histogram arithmetics. #13267
```
Comment (bboreham) [inline CHANGELOG.md:26]: Could bundle together various native histogram optimisations?  #13267, #13215, #13276 #13289, #13340

File: CHANGELOG.md [hunk 17]
```diff
@@ -1,5 +1,37 @@
 # Changelog
 
+## 2.50.0-rc.0 / 2024-01-26
+
+* [FEATURE] Add support for ingesting zeros as created timestamps (enabled under the feature-flag `created-timestamp-zero-ingestion`). #12733 #13279
+* [FEATURE] TSDB/{index,compact}: allow using custom postings encoding format. #13242 #13407 #13391
```
[the remark this answers is not in the document]
Comment (Nexucis) [reply CHANGELOG.md:6]: ok I will remove it. But "user" can mean so many different profiles. Here it could be a developer that would still be interested about this feature don't you think ?

File: CHANGELOG.md [hunk 18]
```diff
@@ -1,5 +1,37 @@
...
+* [FEATURE] Promtool: Add `analyze` histograms command #12331
+* [FEATURE] TSDB/compaction: Add a way to enable overlapping compaction #13282 #13393 #13398
+* [FEATURE] Add automatic memory limit handling. Activated using the feature flag `auto-gomemlimit` #13395
+* [ENHANCEMENT] Promtool: allow setting multiple matchers to `promtool tsdb dump`. #13296
+* [ENHANCEMENT] PromQL: Restore more efficient version of `NewPossibleNonCounterInfo` annotation. #13022
+* [ENHANCEMENT] Enable reusing memory when converting between histogram types. #13215
+* [ENHANCEMENT] PERFORMANCE: Reuse float histogram objects. #13276 #13289
```
[the remark this answers is not in the document]
Comment (Nexucis) [reply CHANGELOG.md:13]: Yeah I know but there are so many PRs relative to the performance, could be interested to introduce it

File: CHANGELOG.md [hunk 19]
```diff
@@ -1,5 +1,37 @@
 # Changelog
 
+## 2.50.0-rc.0 / 2024-01-26
+
+* [FEATURE] Add support for ingesting zeros as created timestamps (enabled under the feature-flag `created-timestamp-zero-ingestion`). #12733 #13279
```
[the remark this answers is not in the document]
Comment (Nexucis) [reply CHANGELOG.md:5]: isn't obvious since it's behind a feature-flag ?

File: CHANGELOG.md (hunk 6 above)
Comment (Nexucis) [reply CHANGELOG.md:32]: I think when I wrote this, I wanted to stamp, it's a change when you are activating it. I'm even wondering if you can rollback after activating it. It's probably the case.

File: CHANGELOG.md (hunk 16 above)
Comment (Nexucis) [reply CHANGELOG.md:26]: yeah that will be more consistant.

File: CHANGELOG.md (hunk 15 above)
Comment (Nexucis) [reply CHANGELOG.md]: done. Let me know if you think it is fuzzy

File: CHANGELOG.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (Nexucis) [reply CHANGELOG.md]: done

File: CHANGELOG.md (hunk 9 above)
Comment (Nexucis) [reply CHANGELOG.md]: well either way is fine for me.
Perhaps I should rephrase it to make it more user oriented like @bboreham is suggesting ?

Comment (Nexucis) [reply CHANGELOG.md]: Would you be fine with that @roidelapluie ?

**Comment by Nexucis** [comment]
as a side note, if we were using semantic commit, that would simplify a lot these process and discussions. And release won't take ages when someone like me (that are less aware about the meaning of every changes happening in Prometheus), would like to be the release shepherd

File: CHANGELOG.md [hunk 20]
```diff
@@ -1,5 +1,32 @@
...
+* [FEATURE] **Experimental**: Add support for ingesting zeros as created timestamps (enabled under the feature-flag `created-timestamp-zero-ingestion`). #12733 #13279
+* [FEATURE] Promtool: Add `analyze` histograms command #12331
+* [FEATURE] TSDB/compaction: Add a way to enable overlapping compaction #13282 #13393 #13398
+* [FEATURE] Add automatic memory limit handling. Activated using the feature flag `auto-gomemlimit` #13395
+* [ENHANCEMENT] Promtool: allow setting multiple matchers to `promtool tsdb dump`. #13296
+* [ENHANCEMENT] PromQL: Restore more efficient version of `NewPossibleNonCounterInfo` annotation. #13022
+* [ENHANCEMENT] Kuma SD: Extend configuration to allow users to specify ClientId. #13278
+* [ENHANCEMENT] PromQL: Use natural sort in `sort_by_label` and `sort_by_label_desc`. This is **experimental**. #13411
+* [ENHANCEMENT] Native Histograms: support `native_histogram_min_bucket_factor` in scrape_config. #13222
+* [ENHANCEMENT] Metrics: Add warnings for histogramRate applied with isCounter not matching counter/gauge histogram. #13392
```
Comment (bboreham) [inline CHANGELOG.md:16]:
```suggestion
* [ENHANCEMENT] Native Histograms: Issue warning if histogramRate is applied to the wrong kind of histogram. #13392
```

(this is the title of the issue, which seems more meaningful to me)

File: CHANGELOG.md [hunk 21]
```diff
@@ -1,5 +1,32 @@
...
+* [ENHANCEMENT] TSDB: Make transaction isolation data structures smaller. #13015
+* [ENHANCEMENT] TSDB/postings: Optimize merge using Loser Tree. #12878
+* [ENHANCEMENT] TSDB: Simplify internal series delete function. #13261
+* [ENHANCEMENT] AGENT: Performance improvement by making the global hash lookup table smaller. #13262
+* [ENHANCEMENT] Labels: Add DropMetricName function, used in PromQL. #13446
```
Comment (bboreham) [inline CHANGELOG.md:21]:
```suggestion
* [ENHANCEMENT] PromQL: faster execution of metric functions, e.g. abs(), rate() #13446
```

(citing the benefit rather than the low-level implementation)

File: CHANGELOG.md [hunk 22]
```diff
@@ -1,5 +1,32 @@
 # Changelog
 
+## 2.50.0-rc.0 / 2024-02-07
+
+* [CHANGE] Remote Write: Drop old inmemory samples. Activated using the config entry `sample_age_limit` #13002
```
Comment (bboreham) [inline CHANGELOG.md:5]: This feels more like a feature than a change.

**Review by bboreham — APPROVED** [review]
Few more suggestions; I'm ok to go as it is though.

**Comment by bboreham** [comment]
As I understand "semantic commits", they do not avoid the problem that PRs describe things from a coder's point of view while release notes are read by non-coders, primarily Prometheus admins.

I'm not sure we even want commit messages to be phrased for non-coders, though I could see having an additional section that gave that version.

**Comment by ArthurSens** [comment]
> as a side note, if we were using semantic commit, that would simplify a lot these process and discussions. And release won't take ages when someone like me (that are less aware about the meaning of every changes happening in Prometheus), would like to be the release shepherd

FYI, @kakkoyun and I are going to experiment Changelog automation in client_golang through LFX mentorship, we can share the experience once it's done :)

File: CHANGELOG.md [hunk 23]
```diff
@@ -1,5 +1,32 @@
 # Changelog
 
+## 2.50.0-rc.0 / 2024-02-07
+
+* [CHANGE] Remote Write: Error `storage.ErrTooOldSample` is now generating HTTP error 400 instead of HTTP error 500. #13335
+* [FEATURE] Remote Write: Drop old inmemory samples. Activated using the config entry `sample_age_limit` #13002
```
Comment (LeviHarrison) [inline CHANGELOG.md:6]: Usually there are periods at the end of entries before the PR references

```suggestion
* [FEATURE] Remote Write: Drop old inmemory samples. Activated using the config entry `sample_age_limit`. #13002
```

File: CHANGELOG.md [hunk 24]
```diff
@@ -1,5 +1,32 @@
...
+* [FEATURE] **Experimental**: Add support for ingesting zeros as created timestamps (enabled under the feature-flag `created-timestamp-zero-ingestion`). #12733 #13279
+* [FEATURE] Promtool: Add `analyze` histograms command #12331
```
Comment (LeviHarrison) [inline CHANGELOG.md:8]:
```suggestion
* [FEATURE] Promtool: Add `analyze` histograms command. #12331
```

File: CHANGELOG.md [hunk 25]
```diff
@@ -1,5 +1,32 @@
...
+* [FEATURE] TSDB/compaction: Add a way to enable overlapping compaction #13282 #13393 #13398
```
Comment (LeviHarrison) [inline CHANGELOG.md:9]:
```suggestion
* [FEATURE] TSDB/compaction: Add a way to enable overlapping compaction. #13282 #13393 #13398
```

File: CHANGELOG.md [hunk 26]
```diff
@@ -1,5 +1,32 @@
...
+* [FEATURE] Add automatic memory limit handling. Activated using the feature flag `auto-gomemlimit` #13395
```
Comment (LeviHarrison) [inline CHANGELOG.md:10]:
```suggestion
* [FEATURE] Add automatic memory limit handling. Activated using the feature flag `auto-gomemlimit`. #13395
```

File: CHANGELOG.md [hunk 27]
```diff
@@ -1,5 +1,32 @@
...
+* [ENHANCEMENT] PromQL: Restore more efficient version of `NewPossibleNonCounterInfo` annotation. #13022
+* [ENHANCEMENT] Kuma SD: Extend configuration to allow users to specify ClientId. #13278
+* [ENHANCEMENT] PromQL: Use natural sort in `sort_by_label` and `sort_by_label_desc`. This is **experimental**. #13411
+* [ENHANCEMENT] Native Histograms: support `native_histogram_min_bucket_factor` in scrape_config. #13222
+* [ENHANCEMENT] Native Histograms: Issue warning if histogramRate is applied to the wrong kind of histogram. #13392
+* [ENHANCEMENT] TSDB: Make transaction isolation data structures smaller. #13015
+* [ENHANCEMENT] TSDB/postings: Optimize merge using Loser Tree. #12878
+* [ENHANCEMENT] TSDB: Simplify internal series delete function. #13261
+* [ENHANCEMENT] AGENT: Performance improvement by making the global hash lookup table smaller. #13262
+* [ENHANCEMENT] PromQL: faster execution of metric functions, e.g. abs(), rate() #13446
```
Comment (LeviHarrison) [inline CHANGELOG.md:21]:
```suggestion
* [ENHANCEMENT] PromQL: faster execution of metric functions, e.g. abs(), rate(). #13446
```

File: CHANGELOG.md [hunk 28]
```diff
@@ -1,5 +1,32 @@
...
 
+## 2.50.0-rc.0 / 2024-02-07
+
+* [CHANGE] Remote Write: Error `storage.ErrTooOldSample` is now generating HTTP error 400 instead of HTTP error 500. #13335
+* [FEATURE] Remote Write: Drop old inmemory samples. Activated using the config entry `sample_age_limit` #13002
+* [FEATURE] **Experimental**: Add support for ingesting zeros as created timestamps (enabled under the feature-flag `created-timestamp-zero-ingestion`). #12733 #13279
+* [FEATURE] Promtool: Add `analyze` histograms command #12331
+* [FEATURE] TSDB/compaction: Add a way to enable overlapping compaction #13282 #13393 #13398
+* [FEATURE] Add automatic memory limit handling. Activated using the feature flag `auto-gomemlimit` #13395
+* [ENHANCEMENT] Promtool: allow setting multiple matchers to `promtool tsdb dump`. #13296
```
Comment (LeviHarrison) [inline CHANGELOG.md:11]: While the above is what the PR title says, I found it a little confusing, especially the use of "to" ("setting" vs "specifying" doesn't really matter)

```suggestion
* [ENHANCEMENT] Promtool: allow specifying multiple matchers in `promtool tsdb dump`. #13296
```

File: CHANGELOG.md [hunk 29]
```diff
@@ -1,5 +1,32 @@
...
+* [ENHANCEMENT] PromQL: Restore more efficient version of `NewPossibleNonCounterInfo` annotation. #13022
+* [ENHANCEMENT] Kuma SD: Extend configuration to allow users to specify ClientId. #13278
```
Comment (LeviHarrison) [inline CHANGELOG.md:13]: Since "ClientID" isn't even the actual field name that got added here this might make more sense.

```suggestion
* [ENHANCEMENT] Kuma SD: Extend configuration to allow users to specify a client ID. #13278
```

File: CHANGELOG.md [hunk 30]
```diff
@@ -1,5 +1,32 @@
...
+* [ENHANCEMENT] PromQL: Use natural sort in `sort_by_label` and `sort_by_label_desc`. This is **experimental**. #13411
+* [ENHANCEMENT] Native Histograms: support `native_histogram_min_bucket_factor` in scrape_config. #13222
+* [ENHANCEMENT] Native Histograms: Issue warning if histogramRate is applied to the wrong kind of histogram. #13392
+* [ENHANCEMENT] TSDB: Make transaction isolation data structures smaller. #13015
+* [ENHANCEMENT] TSDB/postings: Optimize merge using Loser Tree. #12878
+* [ENHANCEMENT] TSDB: Simplify internal series delete function. #13261
```
Comment (LeviHarrison) [inline CHANGELOG.md:19]: This seems like more of an internal thing?

File: CHANGELOG.md [hunk 31]
```diff
@@ -1,5 +1,32 @@
...
+* [ENHANCEMENT] AGENT: Performance improvement by making the global hash lookup table smaller. #13262
```
Comment (LeviHarrison) [inline CHANGELOG.md:20]:
```suggestion
* [ENHANCEMENT] Agent: Performance improvement by making the global hash lookup table smaller. #13262
```

File: CHANGELOG.md [hunk 32]
```diff
@@ -1,5 +1,32 @@
...
+* [ENHANCEMENT] PromQL: faster execution of metric functions, e.g. abs(), rate() #13446
+* [ENHANCEMENT] TSDB: Optimize label values with matchers by taking shortcuts. #13426
+* [ENHANCEMENT] Kubernetes SD: Check preconditions earlier and avoid unnecessary checks or iterations in kube_sd. #13408
+* [ENHANCEMENT] Promtool: Improve visibility for `promtool test rules` with JSON colored formatting. #13342
+* [ENHANCEMENT] Consoles: Exclude iowait and steal from CPU Utilisation. #9593
+* [ENHANCEMENT] Various improvements and optimizations on Native Histogram. #13267, #13215, #13276 #13289, #13340
+* [BUGFIX] Fix quality value in HTTP Accept header. #13313
```
Comment (LeviHarrison) [inline CHANGELOG.md:27]:
```suggestion
* [BUGFIX] Scraping: Fix quality value in HTTP Accept header. #13313
```

File: CHANGELOG.md [hunk 33]
```diff
@@ -1,5 +1,32 @@
...
+* [ENHANCEMENT] TSDB: Make transaction isolation data structures smaller. #13015
+* [ENHANCEMENT] TSDB/postings: Optimize merge using Loser Tree. #12878
+* [ENHANCEMENT] TSDB: Simplify internal series delete function. #13261
+* [ENHANCEMENT] AGENT: Performance improvement by making the global hash lookup table smaller. #13262
+* [ENHANCEMENT] PromQL: faster execution of metric functions, e.g. abs(), rate() #13446
+* [ENHANCEMENT] TSDB: Optimize label values with matchers by taking shortcuts. #13426
+* [ENHANCEMENT] Kubernetes SD: Check preconditions earlier and avoid unnecessary checks or iterations in kube_sd. #13408
+* [ENHANCEMENT] Promtool: Improve visibility for `promtool test rules` with JSON colored formatting. #13342
+* [ENHANCEMENT] Consoles: Exclude iowait and steal from CPU Utilisation. #9593
+* [ENHANCEMENT] Various improvements and optimizations on Native Histogram. #13267, #13215, #13276 #13289, #13340
```
Comment (LeviHarrison) [inline CHANGELOG.md:26]:
```suggestion
* [ENHANCEMENT] Various improvements and optimizations on Native Histograms. #13267, #13215, #13276 #13289, #13340
```

File: CHANGELOG.md (hunk 30 above)
Comment (bboreham) [reply CHANGELOG.md:19]: It's a small optimization.

File: CHANGELOG.md [hunk 34]
```diff
@@ -3,28 +3,28 @@
...
+* [ENHANCEMENT] Agent: Performance improvement by making the global hash lookup table smaller. #13262
 * [ENHANCEMENT] PromQL: faster execution of metric functions, e.g. abs(), rate() #13446
 * [ENHANCEMENT] TSDB: Optimize label values with matchers by taking shortcuts. #13426
 * [ENHANCEMENT] Kubernetes SD: Check preconditions earlier and avoid unnecessary checks or iterations in kube_sd. #13408
 * [ENHANCEMENT] Promtool: Improve visibility for `promtool test rules` with JSON colored formatting. #13342
 * [ENHANCEMENT] Consoles: Exclude iowait and steal from CPU Utilisation. #9593
-* [ENHANCEMENT] Various improvements and optimizations on Native Histogram. #13267, #13215, #13276 #13289, #13340
-* [BUGFIX] Fix quality value in HTTP Accept header. #13313
+* [ENHANCEMENT] Various improvements and optimizations on Native Histograms. #13267, #13215, #13276 #13289, #13340
+* [BUGFIX] Scrapping: Fix quality value in HTTP Accept header. #13313
```
Comment (bboreham) [inline CHANGELOG.md:27]:
```suggestion
* [BUGFIX] Scraping: Fix quality value in HTTP Accept header. #13313
```

**Review by bboreham — APPROVED** [review]
Small typo, but approved so we can get this thing out.

**Comment by bboreham** [comment]
Looking very good:

image

image
image
image

**Comment by Nexucis** [comment]
perfect. I think we can cancel prombench then

**Comment by darshanime** [comment]
Is it too late to include https://github.com/prometheus/prometheus/pull/13578?

**Comment by SuperQ** [comment]
@darshanime That seems like something we can pull in. But I'll leave it up to @Nexucis

**Comment by Nexucis** [comment]
yeah definitely let's add it.

**Comment by bboreham** [comment]
I'm interested to know why #11870 was not mentioned.
